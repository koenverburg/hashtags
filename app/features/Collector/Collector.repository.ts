import Axios, { AxiosInstance, AxiosResponse, AxiosError } from 'axios'
import { IRepository } from '../../base/interfaces/IRepository'
import { Storage } from '@helpers'

interface ICollectorRepository<T> extends IRepository<T> {
  listByCategory(v: string): T[]
}

export class CollectorRepository implements ICollectorRepository<string> {
  constructor() {
    this.storage = new Storage('collector')
    const config = {
      baseURL: process.env.COLLECTOR_API,
    }
    this.http = Axios.create(config)
  }
  private storage: Storage
  private http: AxiosInstance

  public listByCategory(name: string): any {
    // console.log(this.storage.get('tags'))
    return ['', '']
  }

  public list(): any {
    // console.log(this.storage.get('tags'))
    return ['', '']
  }

  public add(entity: {} | [] | string) {
    this.storage.set('tags', entity)
  }
  public delete(entity: any) {
    throw new Error('Method not implemented.')
  }
  public update(entity: any) {
    throw new Error('Method not implemented.')
  }
  public find(id: number) {
    return ''
  }
}
