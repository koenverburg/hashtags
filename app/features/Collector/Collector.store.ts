import { action, computed } from 'mobx'
import { CollectorRepository } from './Collector.repository'
import { Statuses } from '@enums'
import { StoreHelper } from '@helpers'
import { IInitialState } from '../../base/interfaces/IInitialState'

const repository = new CollectorRepository()

export const localInitialState: IInitialState<string> = {
  status: Statuses.IDLE,
  entities: '',
}

export class CollectorStore extends StoreHelper<string> {

  constructor(initialState?: IInitialState<string>) {
    super(initialState)
  }

  @computed get tags() {
    if (!this.entities) return []

    const list = this.entities.trim().split(' ')
    const deduplicatedList = this.deduplicate(list)

    return deduplicatedList
  }

  // private checkForHashtagPrefix() { }

  private deduplicate(list: string[]): string[] {
    const cleanList = list
      .map(item => item.trim())
      .filter(item => item.length > 0)

    const sortedList = cleanList.reduce(
      (accumulator: string[], current: string) =>
        accumulator.includes(current) ? accumulator : [...accumulator, current],
      [],
    )

    return sortedList
  }

  public saveData(entity: string) {
    // repository.add(entity)
    this.setData(entity)
  }

  @action public async fetch () {
    this.status = Statuses.PENDING
    const response = await repository.list()

    if (response) this.setData(response.data)
  }
}
