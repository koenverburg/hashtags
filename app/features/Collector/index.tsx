import React, { Fragment, Component } from 'react'
import { observer, inject } from 'mobx-react'
import { CollectorStore } from './Collector.store'
import { IBaseMobxProps } from '../../base/interfaces/IMobBaseProps'
import { Textarea } from '@components'

interface ICollectorProps extends IBaseMobxProps {
  collector?: CollectorStore
}

@inject('collector')
@observer
export class Collector extends Component<ICollectorProps> {

  private handleChange = (event: Event) => {
    const tags = event.target.value.trim().split(' ')
    const value = this.props.collector!.deduplicate(tags)
    console.log(value)
    this.props.collector!.saveData(value)
  }

  public componentDidMount() {
    const { collector } = this.props
    if (!collector!.any()) collector!.fetch()
  }

  public render() {
    const { children, collector } = this.props
    return (
      <>
        <Textarea handleChange={this.handleChange} />
        {collector!.IsRejected && collector!.error ? collector!.error.message : null}
        {collector!.IsFulfilled && !collector!.any() && 'No Items'}
        {collector!.IsFulfilled && collector!.any() &&
          collector!.entities.map(item => (
            <li key={item.tag}>{item.tag}</li>
          ))
        }
      </>
    )
  }
}

