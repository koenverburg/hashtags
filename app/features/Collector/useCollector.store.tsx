import React, { createContext } from 'react'

import { CollectorStore, localInitialState } from './Collector.store'

const createdContext = createContext(new CollectorStore(localInitialState))

export function useCollectorStore() {
  return React.useContext(createdContext)
}
