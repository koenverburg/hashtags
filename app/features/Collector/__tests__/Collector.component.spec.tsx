/* tslint:disable no-implicit-dependencies */
import React from 'react'
import { shallow } from 'enzyme'
import { Provider } from 'mobx-react'
import { Collector } from '..'
import { CollectorStore } from '../Collector.store'

describe('feature - <Collector />', () => {
  let store: CollectorStore

  beforeAll(() => {
    store = new CollectorStore()
  })

  it('renders', () => {
    const feature = shallow(
      (
        <Provider store={store}>
          <Collector />
        </Provider>
      ),
    )

    expect(feature).toMatchSnapshot()
  })
})
