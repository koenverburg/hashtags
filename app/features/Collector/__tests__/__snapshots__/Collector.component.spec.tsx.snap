// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`feature - <Collector /> renders 1`] = `
<ContextProvider
  value={
    Object {
      "store": CollectorStore {
        "entities": undefined,
        "error": undefined,
        "status": "idle",
      },
    }
  }
>
  <Memo(wrappedComponent) />
</ContextProvider>
`;
